File: qtgql/customscalars/customscalars.cpp
#include "inc/qtgql/customscalars/customscalars.hpp"

namespace qtgql {

const QString &DateTimeScalar::to_qt() {
  if (m_should_update) {
    m_cached_to_qt = m_value.toString("hh:mm (dd.mm.yy)");
    m_should_update = false;
  }
  return m_cached_to_qt;
}

const QString &DateTimeScalar::GRAPHQL_NAME() {
  static const QString ret = "DateTime";
  return ret;
}

void DateTimeScalar::deserialize(const QJsonValue &raw_data) {
  m_value =
      QDateTime::fromString(raw_data.toString(), Qt::DateFormat(Qt::ISODate));
  m_should_update = true;
}

void DecimalScalar::deserialize(const QJsonValue &raw_data) {
  m_value = raw_data.toString();
}

const QString &DecimalScalar::GRAPHQL_NAME() {
  static const QString ret = "Decimal";
  return ret;
}

const QString &DecimalScalar::to_qt() { return m_value; }
}  // namespace qtgql

File: qtgqlcodegen/custom_scalars.py
from qtgqlcodegen.schema.typing import CustomScalarDefinition
from qtgqlcodegen.schema.definitions import CustomScalarMap

# An ISO-8601 encoded datetime.
DateTimeScalarDefinition = CustomScalarDefinition(
    type_name="qtgql::customscalars::DateTimeScalar",
    graphql_name="DateTime",
    deserialized_type="QDateTime",
    type_for_proxy="QString",
    include_path="<qtgql/customscalars/customscalars.hpp>",
)
DateScalarDefinition = CustomScalarDefinition(
    type_name="qtgql::customscalars::DateScalar",
    graphql_name="Date",
    deserialized_type="QDate",
    type_for_proxy="QString",
    include_path="<qtgql/customscalars/customscalars.hpp>",
)
TimeScalarDefinition = CustomScalarDefinition(
    type_name="qtgql::customscalars::TimeScalar",
    graphql_name="Time",
    deserialized_type="QTime",
    type_for_proxy="QString",
    include_path="<qtgql/customscalars/customscalars.hpp>",
)

DecimalScalarDefinition = CustomScalarDefinition(
    type_name="qtgql::customscalars::DecimalScalar",
    graphql_name="Decimal",
    deserialized_type="QString",
    type_for_proxy="QString",
    include_path="<qtgql/customscalars/customscalars.hpp>",
)

CUSTOM_SCALARS: CustomScalarMap = {
    DateTimeScalarDefinition.graphql_name: DateTimeScalarDefinition,
    DateScalarDefinition.graphql_name: DateScalarDefinition,
    TimeScalarDefinition.graphql_name: TimeScalarDefinition,
    DecimalScalarDefinition.graphql_name: DecimalScalarDefinition,
}
